File: tetris.py
import tkinter as tk
import random
shapes = [
    [[1, 1, 1, 1]],                             # I shape
    [[1, 1], [1, 1]],                           # O shape
    [[1, 1, 0], [0, 1, 1], [0, 0, 0]],          # Z shape
    [[0, 1, 1], [1, 1]],                        # S shape
    [[1, 1, 1], [0, 1, 0]],                     # T shape
    [[0, 1, 1], [1, 1, 0]],                     # L shape
    [[1, 1, 1], [0, 0, 1]]                      # J shape
]


class Tetris(tk.Tk):
    def __init__(self):
        super().__init__()
        self.title('Tetris by ruzhila.cn')
        self.canvas = tk.Canvas(self, width=200, height=400, bg='white')
        self.canvas.pack()
        self.board = [[0] * 10 for _ in range(20)]
        self.score, self.game_over = 0, False
        self.new_shape()
        self.bind("<Key>", self.key_press)
        self.after(1000, self.fall)

    def key_press(self, event):
        if self.can_move(0, 1):
            if event.keysym == 'Up':
                self.shape = list(zip(*reversed(self.shape)))
            elif event.keysym == 'Down':
                self.shape_y += 1
        if event.keysym == 'Left' and self.can_move(-1, 0):
            self.shape_x -= 1
        elif event.keysym == 'Right' and self.can_move(1, 0):
            self.shape_x += 1
        self.redraw()

    def new_shape(self):
        self.shape = random.choice(shapes)
        self.shape_x = 5
        self.shape_y = 0
        self.color = random.choice(['blue', 'green', 'yellow', 'pink'])

    def fall(self):
        self.game_over = not self.can_move(0, 0)
        if not self.game_over:
            if self.can_move(0, 1):
                self.shape_y += 1
            else:
                self.add_to_board()
                self.check_lines()
                self.new_shape()
            self.after(1000, self.fall)
        self.redraw()

    def can_move(self, dx, dy):
        for i, row in enumerate(self.shape):
            for j, cell in enumerate(row):
                if cell:
                    x = self.shape_x + j + dx
                    y = self.shape_y + i + dy
                    if x < 0 or x >= 10 or y < 0 or y >= 20 or self.board[y][x]:
                        return False
        return True

    def add_to_board(self):
        for i, row in enumerate(self.shape):
            for j, cell in enumerate(row):
                if cell:
                    self.board[self.shape_y + i][self.shape_x + j] = 1

    def check_lines(self):
        for i, row in enumerate(self.board):
            if all(cell for cell in row):
                del self.board[i]
                self.board.insert(0, [0] * 10)
                self.score += 1

    def draw_shape(self, shape, shape_x, shape_y, color):
        for i, row in enumerate(shape):
            for j, cell in enumerate(row):
                if cell:
                    x = (shape_x + j) * 20
                    y = (shape_y + i) * 20
                    self.canvas.create_rectangle(
                        x, y, x + 20, y + 20, fill=color, outline='gray')

    def redraw(self):
        self.canvas.delete('all')
        self.draw_shape(self.board, 0, 0, 'lightgray')
        self.draw_shape(self.shape, self.shape_x, self.shape_y, self.color)
        if self.game_over:
            self.canvas.create_text(100, 200, anchor='center',
                                    text='Game Over', fill='red', font=('Arial', 20))
        self.canvas.create_text(10, 10, anchor='nw',
                                text=f'Score: {self.score} | ruzhila.cn', fill='blue')


if __name__ == '__main__':
    Tetris().mainloop()

File: Tetris.java
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.util.Random;

import javax.swing.JFrame;
import javax.swing.JPanel;
import javax.swing.Timer;

public class Tetris extends JPanel implements ActionListener {
    private static final int SCREEN_WIDTH = 200;
    private static final int SCREEN_HEIGHT = 400;
    private static final int UNIT_SIZE = 20;
    private static final int[][][] shapes = {
            {
                    { 1, 1, 1, 1 }
            },
            {
                    { 1, 1, 1 },
                    { 0, 1, 0 }
            },
            {
                    { 1, 1, 1 },
                    { 1, 0, 0 }
            },
            {
                    { 1, 0 },
                    { 1, 0 },
                    { 1, 1 },
            },
            {
                    { 1, 1, 0 },
                    { 0, 1, 1 }
            },
            {
                    { 1, 1 },
                    { 1, 1 }
            },
            {
                    { 1, 1, 1 },
                    { 0, 0, 1 }
            },
            {
                    { 0, 1, 0 },
                    { 0, 1, 1 },
                    { 0, 1, 0 },
            },
    };
    private Color[] colors = {
            Color.RED, Color.GREEN, Color.ORANGE, Color.CYAN, Color.MAGENTA
    };
    private int[][] board = new int[UNIT_SIZE][SCREEN_WIDTH / UNIT_SIZE];
    private int[][] shape;
    private int shapeX = 0;
    private int shapeY = 0;
    private int score = 0;
    private boolean gameOver = false;
    private Color shapeColor;
    private Timer timer;

    Tetris() {
        setPreferredSize(new Dimension(SCREEN_WIDTH, SCREEN_HEIGHT));
        setVisible(true);
        setFocusable(true);
        newShape();
        timer = new Timer(500, this);
        timer.start();
        addKeyListener(new TetrisKeyAdapter());
    }

    void newShape() {
        shape = shapes[new Random().nextInt(shapes.length)];
        shapeX = 5;
        shapeY = 0;
        shapeColor = colors[new Random().nextInt(colors.length)];
    }

    void rotateShape() {
        int newShape[][] = new int[shape[0].length][shape.length];
        for (int i = 0; i < shape.length; i++) {
            for (int j = 0; j < shape[i].length; j++) {
                newShape[j][shape.length - 1 - i] = shape[i][j];
            }
        }
        shape = newShape;
    }

    void moveShape(int dx, int dy) {
        if (canMove(dx, dy)) {
            shapeX += dx;
            shapeY += dy;
        }
    }

    boolean canMove(int dx, int dy) {
        for (int i = 0; i < shape.length; i++) {
            for (int j = 0; j < shape[i].length; j++) {
                if (shape[i][j] != 1) {
                    continue;
                }
                int newX = shapeX + j + dx;
                int newY = shapeY + i + dy;
                if (newX < 0 || newX >= SCREEN_WIDTH / UNIT_SIZE ||
                        newY >= UNIT_SIZE ||
                        board[newY][newX] == 1) {
                    return false;
                }
            }
        }
        return true;
    }

    void checkLines() {
        for (int i = board.length - 1; i >= 0; i--) {
            boolean isLine = true;
            for (int j = 0; j < board[i].length; j++) {
                if (board[i][j] != 1) {
                    isLine = false;
                    break;
                }
            }
            if (isLine) {
                for (int k = i; k > 0; k--) {
                    for (int j = 0; j < board[k].length; j++) {
                        board[k][j] = board[k - 1][j];
                    }
                }
                i++;
                score += 100;
            }
        }
    }

    class TetrisKeyAdapter extends KeyAdapter {
        public void keyPressed(KeyEvent e) {
            switch (e.getKeyCode()) {
                case KeyEvent.VK_LEFT:
                    moveShape(-1, 0);
                    break;
                case KeyEvent.VK_RIGHT:
                    moveShape(1, 0);
                    break;
                case KeyEvent.VK_DOWN:
                    moveShape(0, 1);
                    break;
                case KeyEvent.VK_UP:
                    rotateShape();
                    break;
            }
            repaint();
        }
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        if (gameOver) {
            return;
        }
        if (canMove(0, 1)) {
            shapeY++;
        } else {
            for (int i = 0; i < shape.length; i++) {
                for (int j = 0; j < shape[i].length; j++) {
                    if (shape[i][j] == 1) {
                        board[shapeY + i][shapeX + j] = 1;
                    }
                }
            }
            checkLines();
            newShape();

            if (!canMove(0, 0)) {
                gameOver = true;
                timer.stop();
            }
        }
        repaint();
    }

    public void paint(Graphics g) {
        g.setColor(Color.WHITE);
        g.fillRect(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);

        for (int i = 0; i < UNIT_SIZE; i++) {
            for (int j = 0; j < SCREEN_WIDTH / UNIT_SIZE; j++) {
                if (board[i][j] == 1) {
                    g.setColor(Color.BLUE);
                    g.fillRect(j * UNIT_SIZE, i * UNIT_SIZE, UNIT_SIZE, UNIT_SIZE);
                    g.setColor(Color.LIGHT_GRAY);
                    g.drawRect(j * UNIT_SIZE, i * UNIT_SIZE, UNIT_SIZE, UNIT_SIZE);
                }
            }
        }

        for (int i = 0; i < shape.length; i++) {
            for (int j = 0; j < shape[i].length; j++) {
                if (shape[i][j] == 1) {
                    g.setColor(shapeColor);
                    g.fillRect((j + shapeX) * UNIT_SIZE, (i + shapeY) * UNIT_SIZE, UNIT_SIZE, UNIT_SIZE);
                    g.setColor(Color.LIGHT_GRAY);
                    g.drawRect((j + shapeX) * UNIT_SIZE, (i + shapeY) * UNIT_SIZE, UNIT_SIZE, UNIT_SIZE);
                }
            }
        }
        g.setColor(Color.BLACK);
        g.drawString("Score: " + score + " by ruzhila.cn", 0, 10);
        if (gameOver) {
            g.setColor(Color.RED);
            g.drawString("Game Over", 50, 50);
        }
    }

    public static void main(String[] args) {
        JFrame frame = new JFrame("Tetris");
        frame.add(new Tetris());
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.pack();
        frame.setVisible(true);
    }
}